### ISSUE TITLE
trying to st_write() to an FGB file using append = TRUE crashes R

### ISSUE THREAD

**Issue by markolipka** [r-spatial/sf#2012]
**Describe the bug**

`st_write(obj = p, dsn = "p.fgb", append = TRUE)` crashes R / bombs RStudio.

I'm not sure if it is even possible to append to a FlatGeobuf file. However, attempting to do so should not lead to a crash.

**To Reproduce**

```
library(sf)

p <- st_point(c(1, 2)) |> st_sfc()

st_write(obj = p, dsn = "p.fgb")
st_write(obj = p, dsn = "p.fgb", append = TRUE)
```

<details>

```
sessionInfo()
R version 4.1.3 (2022-03-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=German_Germany.1252  LC_CTYPE=German_Germany.1252    LC_MONETARY=German_Germany.1252 LC_NUMERIC=C                   
[5] LC_TIME=German_Germany.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] rstudioapi_0.14  knitr_1.40       magrittr_2.0.3   R6_2.5.1         rlang_1.0.6      fastmap_1.1.0    fansi_1.0.3     
 [8] highr_0.9        tools_4.1.3      xfun_0.33        utf8_1.2.2       cli_3.4.1        clipr_0.8.0      withr_2.5.0     
[15] htmltools_0.5.3  yaml_2.3.5       readxl_1.4.1     digest_0.6.29    tibble_3.1.8     lifecycle_1.0.3  processx_3.7.0  
[22] callr_3.7.2      ps_1.7.1         vctrs_0.4.2      fs_1.5.2         glue_1.6.2       evaluate_0.17    rmarkdown_2.17  
[29] reprex_2.0.2     compiler_4.1.3   pillar_1.8.1     cellranger_1.1.0 generics_0.1.3   lubridate_1.8.0  pkgconfig_2.0.3 
```

```
sf::sf_extSoftVersion()
          GEOS           GDAL         proj.4 GDAL_with_GEOS     USE_PROJ_H           PROJ 
      "3.10.2"        "3.4.1"        "7.2.1"         "true"         "true"        "7.2.1" 
```

</details>

**Comment by rsbivand** [comment]
Please do update to the most recent versions of R and the forthcoming Rtools42 (available for R-4.2.1 patched or R-devel https://cran.r-project.org/bin/windows/Rtools/rtools42/files/rtools42-5355-5357.exe). There is no point resolving issues in non-current releases that may already be fixed upstream or in later releases. Why: Fedora 36, R 4.2.1, **sf** development version, command line:
```
> library(sf)
Linking to GEOS 3.11.0, GDAL 3.5.2, PROJ 9.1.0; sf_use_s2() is TRUE
> p.fgb <- tempfile(fileext=".fgb")
> p <- st_point(c(1, 2)) |> st_sfc()
> st_write(obj = p, dsn = p.fgb)
Writing layer `file373c53e1ba853' to data source 
  `/tmp/RtmpLM1JP0/file373c53e1ba853.fgb' using driver `FlatGeobuf'
Writing 1 features with 0 fields and geometry type Point.
> st_write(obj = p, dsn = p.fgb, append=TRUE)
Updating layer `file373c53e1ba853' to data source `/tmp/RtmpLM1JP0/file373c53e1ba853.fgb' using driver `FlatGeobuf'
Updating existing layer file373c53e1ba853
Writing 1 features with 0 fields and geometry type Point.
> st_read(p.fgb)
Reading layer `file373c53e1ba853' from data source 
  `/tmp/RtmpLM1JP0/file373c53e1ba853.fgb' using driver `FlatGeobuf'
Simple feature collection with 2 features and 0 fields
Geometry type: POINT
Dimension:     XY
Bounding box:  xmin: 1 ymin: 2 xmax: 1 ymax: 2
CRS:           NA
```
I can check on Windows if need be, but suspect GDAL https://github.com/OSGeo/gdal/issues/5739 is connected. This you could have checked.

**Comment by edzer** [comment]
Thanks, confirmed!
